enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,423 @@
+
+End users will have 2 possible variations when using this functionality,
+
+- Create a route for the user workload and this is completely managed by the end user.
+- Create an ingress for the user workload and a managed route is created automatically
+  by the ingress-to-route controller.
+
+Both these workflows will support integrating with third party certificate management
+solutions like cert-manager with the annotation described below.
+
+**When a Route is directly used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Route is going to be created.
+- To expose a user workload, the user would create a new Route with an annotation
+  referencing the generated secret that was created in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair generated, the secret-injector controller copies the certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+**When an Ingress is used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Ingress is going to be created.
+- To expose a user workload, he would need to create a new Ingress with the generated secret
+  referenced in `.spec.tls.secretName`.
+- If the secret CR that is referenced exists and has a successfully generated
+  cert/key pair, the ingress-to-route controller copies this certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+_Note_: The same workflow would apply to updating an existing Route to integrate
+with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
+and `.spec.tls.key` in the route CR since the annotation gets a higher preference.
+
+#### Certificate Life cycle
+
+Post integrating with third-party solutions like cert-manager, the end user can also
+disable automatic certificate management. This can be done in 2 ways,
+
+- The end user can delete the annotation on the Route.
+- The end user deletes the secret that was associated with the Certificate CR.
+
+**Deleting the annotation on the Route**
+The end user can delete the annotation on the Route, this will result
+in abondoning the cert/key pair that was generated via cert-manager and resort to
+using the default certificates that were initially generated by the cluster-ingress-operator
+i.e. clearing the `.spec.tls` on Route.
+
+**Deleting the generated secret that contains the serving cert/key pair**
```

Comment:
1. Is the secret deletion behavior similar in case of Ingress? 
2. What'd be the usecase here for user to delete secrets? In case of compromised secrets, it'd be more useful for user to manual override by removing annotation or update certificate chain and keep the same secret. The secret is believed to be bearing the new certificates. 

In case certificate management is removed, it is better to expect user intervention to update the annotation as it'd an expected behavior for user.

Reply:
> Is the secret deletion behavior similar in case of Ingress?

The ingress-to-route controller, doesn't create a route when the secret is not found. Additional details are [here](https://github.com/openshift/enhancements/pull/1307#discussion_r1073254708).

> What'd be the usecase here for user to delete secrets?

Mostly just covering all bases (what ifs and other unintended actions).